frontends/ast/simplify.cc

```diff
@@ -1902,89 +1865,70 @@ bool AstNode::simplify(bool const_fold, int stage, int width_hint, bool sign_hin
 
-			// Insert clones children from template at beginning
-			for (int i  = 0; i < GetSize(template_node->children); i++)
-				children.insert(children.begin() + i, template_node->children[i]->clone());
+			// Add packed dimensions.
+			if (add_packed_dimensions) {
+				AstNode *packed = children[1];
+				if (newNode->children.empty())
+					newNode->children.insert(newNode->children.begin(), packed->clone());
+				else
+					prepend_ranges(newNode->children[0], packed);
+			}
 
-			if (type == AST_MEMORY && GetSize(children) == 1) {
-				// Single-bit memories must have [0:0] range
-				AstNode *rng = make_range(0, 0);
-				children.insert(children.begin(), rng);
+			// Add unpacked dimensions.
+			if (type == AST_MEMORY) {
+				AstNode *unpacked = children.back();
+				if (GetSize(newNode->children) < 2)
+					newNode->children.push_back(unpacked->clone());
+				else
+					prepend_ranges(newNode->children[1], unpacked);
+				newNode->type = type;
 			}
-			fixup_hierarchy_flags();
-			did_something = true;
+
+			// Prepare to generate dimensions metadata for the resolved type.
+			newNode->dimensions.clear();
+			newNode->unpacked_dimensions = 0;
+
+			goto apply_newNode;
 		}
-		log_assert(!is_custom_type);
 	}
 
 	// resolve types of parameters
 	if (type == AST_LOCALPARAM || type == AST_PARAMETER) {
 		if (is_custom_type) {
-			log_assert(children.size() == 2);
+			log_assert(children.size() >= 2);
 			log_assert(children[1]->type == AST_WIRETYPE);
-			auto type_name = children[1]->str;
-			if (!current_scope.count(type_name)) {
-				input_error("Unknown identifier `%s' used as type name\n", type_name.c_str());
-			}
-			AstNode *resolved_type_node = current_scope.at(type_name);
-			if (resolved_type_node->type != AST_TYPEDEF)
-				input_error("`%s' does not name a type\n", type_name.c_str());
-			log_assert(resolved_type_node->children.size() == 1);
-			AstNode *template_node = resolved_type_node->children[0];
-
-			// Ensure typedef itself is fully simplified
-			while (template_node->simplify(const_fold, stage, width_hint, sign_hint)) {};
 
-			if (template_node->type == AST_STRUCT || template_node->type == AST_UNION) {
-				// replace with wire representing the packed structure
-				newNode = make_packed_struct(template_node, str, attributes);
-				newNode->set_attribute(ID::wiretype, mkconst_str(resolved_type_node->str));
-				newNode->type = type;
-				current_scope[str] = this;
-				// copy param value, it needs to be 1st value
-				delete children[1];
-				children.pop_back();
-				newNode->children.insert(newNode->children.begin(), children[0]->clone());
-				goto apply_newNode;
-			}
-			delete children[1];
-			children.pop_back();
+			// Pretend it's just a wire in order to resolve the type in the code block above.
+			AstNodeType param_type = type;
+			type = AST_WIRE;
+			AstNode *expr = children[0];
+			children.erase(children.begin());
+			while (is_custom_type && simplify(const_fold, stage, width_hint, sign_hint)) {};
```

Comment:
If this type substitution pattern proliferates, let's factor it out. You can leave it for now.

Reply:
OK, I'll keep it in mind.